### ISSUE TITLE
Save/Load corruption bug for native Linux

### ISSUE THREAD

**Issue by bryanperris** [kevinbentley/Descent3#7]
How to reproduce:
1. Install the game from Steam on a Linux based system (Steam Deck for me)
2. Start a new game and make a save
3. Attempt to load the save, sometimes it will be fine.
4. If the game saves a bad save file, it crashes the game upon load.

**Comment by kevinbentley** [comment]
Can you verify that this bug still exists once the source has been compiled? We fixed a bunch of bugs that never made it into release. I'll try to find the list we had.

**Comment by bryanperris** [comment]
Do you have the official build instructions I should follow for Linux? Once its built, I can make a duplicate of the D3 Steam directory, and replace the binary executable with the GitHub version to test on the Steam Deck.

**Comment by bryanperris** [comment]
I think for my reported tickets, there should be a Steam label just to track, I found this issue from Steam version and not from this repo.

**Comment by bryanperris** [comment]
Agreed, I meant more that the label to mark 'bug' issues not being related to this repo, I have no doubt others will also post here bugs they found on Steam D3 version.

**Comment by JeodC** [comment]
> Author

I can see how that could cause confusion. Label created and applied.

**Comment by bryanperris** [comment]
This issue does exist

So in the terminal a lot of these messages
RECOMPILE YOUR SCRIPTS!!!
Game-Checksum FAIL!!! (2273889867l!=2273873307l)
RECOMPILE YOUR SCRIPTS!!!
Game-Checksum FAIL!!! (2273889867l!=2273873307l)
RECOMPILE YOUR SCRIPTS!!!
Game-Checksum FAIL!!! (2273889867l!=2273873307l)
RECOMPILE YOUR SCRIPTS!!!
Game-Checksum FAIL!!! (2273889867l!=2273873307l)
RECOMPILE YOUR SCRIPTS!!!

Then what I caught in GDB:
`Thread 1 "Descent3" received signal SIGSEGV, Segmentation fault.
0x000000000053ae1a in ObjUnlink (objnum=236) at /home/bperris/base/dev/Descent3/Descent3/object.cpp:1478
1478	      seg->objects = obj->next;
`

**Comment by bryanperris** [comment]
@kevinbentley 
I have applied this change to object.cpp line 1475:
terrain_segment *seg = &Terrain_seg[cellnum & 0xFFFF];

No more crash, my save file loads, ship restores to the right place, gameplay seems normal, I am not sure how to verify all the information without some kind of compare log between memory states.  Poking you, if it make sense that cellnum should be wrapping around on the terrain table.

**Comment by JeodC** [comment]
> @kevinbentley I have applied this change to object.cpp line 1475: terrain_segment *seg = &Terrain_seg[cellnum & 0xFFFF];
> 
> No more crash, my save file loads, ship restores to the right place, gameplay seems normal, I am not sure how to verify all the information without some kind of compare log between memory states. Poking you, if it make sense that cellnum should be wrapping around on the terrain table.
> 
> Edit: so I noticed the terrain segment isn't null, but has some non-zero memory address, then printed the cellnum value, and noticed it way beyond the table's memory range, so no longer it was crashing. I would like to investigate more on why the cellnum went beyond the range anyways.
> 
> I found the cellnum mask: #define ROOMNUM_CELLNUM_MASK 0x7fffffff This signed 32-bit mask is way larger than 16 bits.

Is this resolved or just partially resolved?

**Comment by bryanperris** [comment]
I don't feel it is resolved, only found a simple way to avoid the crash.  It is going to take time to figure out a better solution since it doesn't affect Windows builds.

**Comment by bryanperris** [comment]
closer to the issue here:
when load routines reads in the objects, it is reading in a negative value

>>>> Read roomnum: -2147447883
>>>> Read roomnum: -2147447883

**Comment by bryanperris** [comment]
here is a dump of read room numbers from the object file
https://pastebin.com/p8qVPSJ9

**Comment by bryanperris** [comment]
I applied the 16-bit mask to all of the terrain cell calls, but I cannot open the first door in level 2 after loading my problematic save file.

**Comment by viggen66** [comment]
Can you share your compiled Descent3 binary with saved game fixed?

**Comment by bryanperris** [comment]
I have not fixed the issue but only prevent crashing with a hack, still investigating.
